### PR TITLE
Check deathtiles in core movement

### PR DESCRIPTION
**Pull request by AssassinTee**
Addressed issue #1936 and fixed it with a digital differential analyzer for best performance

[https://en.wikipedia.org/wiki/Digital_differential_analyzer_(graphics_algorithm)](https://en.wikipedia.org/wiki/Digital_differential_analyzer_(graphics_algorithm))

We should talk about this before merging, since I think you may have some conerns

I tested it with a falling testmap and/or gravity 500, gravity 999 was to high.

### CODE REVIEW

**Comment by AssassinTee** [comment]
I don't understand why it needs to be GetProximityRadius/2

File: src/game/gamecore.cpp [hunk 1]
```diff
@@ -388,7 +389,15 @@ void CCharacterCore::Move()
 		}
 	}
 
-	m_Pos = NewPos;
+	// check player death by deathtiles
+	vec2 LastDeathPos;
+	if(m_pCollision->HitTileDeath(m_Pos, NewPos, LastDeathPos, 28.0f/2))//Gets tee radius
```
Comment (Dune-jr) [inline src/game/gamecore.cpp]: Indent issues around here

File: src/game/server/entities/character.cpp [hunk 2]
```diff
@@ -529,10 +529,7 @@ void CCharacter::Tick()
 	m_Core.Tick(true);
 
 	// handle death-tiles and leaving gamelayer
-	if(GameServer()->Collision()->GetCollisionAt(m_Pos.x+GetProximityRadius()/3.f, m_Pos.y-GetProximityRadius()/3.f)&CCollision::COLFLAG_DEATH ||
-		GameServer()->Collision()->GetCollisionAt(m_Pos.x+GetProximityRadius()/3.f, m_Pos.y+GetProximityRadius()/3.f)&CCollision::COLFLAG_DEATH ||
-		GameServer()->Collision()->GetCollisionAt(m_Pos.x-GetProximityRadius()/3.f, m_Pos.y-GetProximityRadius()/3.f)&CCollision::COLFLAG_DEATH ||
-		GameServer()->Collision()->GetCollisionAt(m_Pos.x-GetProximityRadius()/3.f, m_Pos.y+GetProximityRadius()/3.f)&CCollision::COLFLAG_DEATH ||
+	if(m_Core.m_DiesFromDeathtiles ||
```
Comment (Dune-jr) [inline src/game/server/entities/character.cpp]: Couldn't we simply call `GameServer()->Collision()->HitTileDeath(...)` here? This way the order in which the `Tick` methods are called is not important.

File: src/game/collision.cpp [hunk 3]
```diff
@@ -202,3 +202,59 @@ void CCollision::MoveBox(vec2 *pInoutPos, vec2 *pInoutVel, vec2 Size, float Elas
...
+
+    dx = dx / step;
+    dy = dy / step;
+
+    float x = Pos1.x;
+    float y = Pos1.y;
+
+    int i = 1;
+
+    while(i <= step) {
```
Comment (Dune-jr) [inline src/game/collision.cpp]: Teeworlds uses 
```
while
{
}
```
style brackets

File: src/game/collision.cpp [hunk 4]
```diff
@@ -202,3 +202,59 @@ void CCollision::MoveBox(vec2 *pInoutPos, vec2 *pInoutVel, vec2 Size, float Elas
 	*pInoutPos = Pos;
 	*pInoutVel = Vel;
 }
+
+bool CCollision::HitTileDeath(vec2& Pos, vec2& Newpos, vec2& Deathpos, float radius)
+{
+    vec2 dir, orth, deathpos1, deathpos2;
```
Comment (Dune-jr) [inline src/game/collision.cpp]: Those variables need to be capitalized

File: src/game/collision.cpp [hunk 5]
```diff
@@ -202,3 +202,59 @@ void CCollision::MoveBox(vec2 *pInoutPos, vec2 *pInoutVel, vec2 Size, float Elas
...
+    return false;
+}
+
+// finds any deathtile between pos1 and pos2 using digital differential analyser
+bool CCollision::CheckDeath(vec2 Pos1, vec2 Pos2, vec2& Deathpos)
+{
+    float dx = (Pos2.x - Pos1.x);
+    float dy = (Pos2.y - Pos1.y);
+
+    float step;
```
Comment (Dune-jr) [inline src/game/collision.cpp]: Step

File: src/game/server/entities/character.cpp (hunk 2 above)
Comment (AssassinTee) [reply src/game/server/entities/character.cpp]: For Hit Tile Death we need to know the Future position, ofc you could do that there

File: src/game/collision.cpp (hunk 3 above)
Comment (AssassinTee) [reply src/game/collision.cpp]: Nice to know 👍

File: src/game/gamecore.cpp (hunk 1 above)
Comment (AssassinTee) [reply src/game/gamecore.cpp]: I can't find any oO pls explain

File: src/game/server/entities/character.cpp (hunk 2 above)
Comment (AssassinTee) [reply src/game/server/entities/character.cpp]: But the calculation for this is in gamecore move()

File: src/game/gamecore.cpp (hunk 1 above)
Comment (Dune-jr) [reply src/game/gamecore.cpp]: image

Probably a mix of spaces and tabs

File: src/game/collision.cpp (hunk 3 above)
Comment (Dune-jr) [reply src/game/collision.cpp]: Also applies to the if below ;)

File: src/game/gamecore.cpp (hunk 1 above)
Comment (AssassinTee) [reply src/game/gamecore.cpp]: probably a problem from my IDE, i will take a look

Comment (AssassinTee) [reply src/game/gamecore.cpp]: is the intended way tabs or spaces?

File: src/game/collision.cpp [hunk 6]
```diff
@@ -202,3 +202,60 @@ void CCollision::MoveBox(vec2 *pInoutPos, vec2 *pInoutVel, vec2 Size, float Elas
...
+        if(GetCollisionAt(x, y)&COLFLAG_DEATH) {
+            Deathpos = vec2(x, y);
+            return true;
+        }
+        x = x + dx;
+        y = y + dy;
+
+        ++i;
+    }
+    Deathpos = vec2(x, y);
```
Comment (Dune-jr) [inline src/game/collision.cpp]: I don't understand why `Deathpos` should be set here, when no collision with a death tile happened. Imo when only one `CheckDeath` returns true, `HitTileDeath` should not choose the closer death position but simply the only actual death position. Am I missing something?

File: src/game/gamecore.cpp (hunk 1 above)
Comment (Dune-jr) [reply src/game/gamecore.cpp]: Tabs

Comment (AssassinTee) [reply src/game/gamecore.cpp]: fixed

File: src/game/collision.cpp (hunk 6 above)
Comment (AssassinTee) [reply src/game/collision.cpp]: Line 259 is simply there, that Deathpos contains valid information, we could simply delete it

I didn't know you actually had a written code style guide. I will work through that, but the HitTileDeath method should look another way

**Comment by AssassinTee** [comment]
How does this code work?

- Checks on position update, if a line between the old and the new position intersects a deathtile
- Is it one line? No, there are two, one left and one right of the Tee when falling down
- the lines are parallel to the movement direction
- the distance of the lines to the actual movement line is exactly the Tees radius; calculate orthogonal line to movement direction, normalize it und multiply by tee radius
- One Line does hit a deathtile and the other not? Doesn't matter, player dies
- The deathposition is not perfectly calculated, but very close (should I do that?)

File: src/game/collision.cpp (hunk 6 above)
Comment (Dune-jr) [reply src/game/collision.cpp]: > I didn't know you actually had a written code style guide.

I didn't write any of that, almost all the documentation is written by previous developers, probably matricks. It's not necessarily my favorite code style either, it's just what Teeworlds uses and inherits.

The documentation's github lives there btw: https://github.com/teeworlds/documentation

**Comment by nheir** [comment]
Could a variation of MoveBox (that is step by step) be used to move tee and stop it on a death tile instead of doing it afterwards ?

DDA is nice, it could also be used for IntersectLine ( https://github.com/nheir/teeworlds/compare/fast_intersectline )

**Comment by AssassinTee** [comment]
I noticed, that my code does not get the exact death position.
At least i get it 100% right in the cases a player should die or not

**Comment by AssassinTee** [comment]
I noticed a huge problem:
If you would hit another player first (player player collision), this prediction doesn't work.
Otherwise I added a few things now:
- Exact death position calculation
- Line-Line Intersection calculation

**Pull request closed without merging**